# CPSC 483-01 Project 6 Imbalanced Dataset Influence 5/6/2021
   ## Student Name: Fulya Kocaman
   ## Student ID: 803023878
   
 ## Project 6 Description:

In this project I used a real-world dataset called Bank Marketing Dataset from UCI and observed hoe an imbalanced dataset can influence classifier performance. In addition to the libraries I have used in previous projects, I used imbalanced-learn to do random oversampling.

I implemented all of the required parts of this project and answered all of the experiment questions in the markdown cells of my code below. As a conclusion, I can say that my code seems to be working fine and predicting accurate results to the given experiment questions.

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.naive_bayes import GaussianNB
from sklearn.neighbors import KNeighborsClassifier
from sklearn.pipeline import make_pipeline
from sklearn.svm import SVC

## Experiments 1 and 2: 
Download the bank-additional.csv for training and bank-additional-full.csv for testing, and then examine each dataset.

In [2]:
dfTrainingSet = pd.read_csv('bank-additional.csv', sep=';')
dfTrainingSet

age          job  marital          education default  housing     loan  \
0      30  blue-collar  married           basic.9y      no      yes       no   
1      39     services   single        high.school      no       no       no   
2      25     services  married        high.school      no      yes       no   
3      38     services  married           basic.9y      no  unknown  unknown   
4      47       admin.  married  university.degree      no      yes       no   
...   ...          ...      ...                ...     ...      ...      ...   
4114   30       admin.  married           basic.6y      no      yes      yes   
4115   39       admin.  married        high.school      no      yes       no   
4116   27      student   single        high.school      no       no       no   
4117   58       admin.  married        high.school      no       no       no   
4118   34   management   single        high.school      no      yes       no   

        contact month day_of_week  ...  campaign  pdays  previous  \
0      cellular   may         fri  ...         2    999         0   
1     telephone   may         fri  ...         4    999         0   
2     telephone   jun         wed  ...         1    999         0   
3     telephone   jun         fri  ...         3    999         0   
4      cellular   nov         mon  ...         1    999         0   
...         ...   ...         ...  ...       ...    ...       ...   
4114   cellular   jul         thu  ...         1    999         0   
4115  telephone   jul         fri  ...         1    999         0   
4116   cellular   may         mon  ...         2    999         1   
4117   cellular   aug         fri  ...         1    999         0   
4118   cellular   nov         wed  ...         1    999         0   

         poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
0     nonexistent         -1.8          92.893          -46.2      1.313   
1     nonexistent          1.1          93.994          -36.4      4.855   
2     nonexistent          1.4          94.465          -41.8      4.962   
3     nonexistent          1.4          94.465          -41.8      4.959   
4     nonexistent         -0.1          93.200          -42.0      4.191   
...           ...          ...             ...            ...        ...   
4114  nonexistent          1.4          93.918          -42.7      4.958   
4115  nonexistent          1.4          93.918          -42.7      4.959   
4116      failure         -1.8          92.893          -46.2      1.354   
4117  nonexistent          1.4          93.444          -36.1      4.966   
4118  nonexistent         -0.1          93.200          -42.0      4.120   

      nr.employed   y  
0          5099.1  no  
1          5191.0  no  
2          5228.1  no  
3          5228.1  no  
4          5195.8  no  
...           ...  ..  
4114       5228.1  no  
4115       5228.1  no  
4116       5099.1  no  
4117       5228.1  no  
4118       5195.8  no  

[4119 rows x 21 columns]

In [3]:
dfTestSet = pd.read_csv('bank-additional-full.csv', sep=';')
dfTestSet

age          job  marital            education  default housing loan  \
0       56    housemaid  married             basic.4y       no      no   no   
1       57     services  married          high.school  unknown      no   no   
2       37     services  married          high.school       no     yes   no   
3       40       admin.  married             basic.6y       no      no   no   
4       56     services  married          high.school       no      no  yes   
...    ...          ...      ...                  ...      ...     ...  ...   
41183   73      retired  married  professional.course       no     yes   no   
41184   46  blue-collar  married  professional.course       no      no   no   
41185   56      retired  married    university.degree       no     yes   no   
41186   44   technician  married  professional.course       no      no   no   
41187   74      retired  married  professional.course       no     yes   no   

         contact month day_of_week  ...  campaign  pdays  previous  \
0      telephone   may         mon  ...         1    999         0   
1      telephone   may         mon  ...         1    999         0   
2      telephone   may         mon  ...         1    999         0   
3      telephone   may         mon  ...         1    999         0   
4      telephone   may         mon  ...         1    999         0   
...          ...   ...         ...  ...       ...    ...       ...   
41183   cellular   nov         fri  ...         1    999         0   
41184   cellular   nov         fri  ...         1    999         0   
41185   cellular   nov         fri  ...         2    999         0   
41186   cellular   nov         fri  ...         1    999         0   
41187   cellular   nov         fri  ...         3    999         1   

          poutcome emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  \
0      nonexistent          1.1          93.994          -36.4      4.857   
1      nonexistent          1.1          93.994          -36.4      4.857   
2      nonexistent          1.1          93.994          -36.4      4.857   
3      nonexistent          1.1          93.994          -36.4      4.857   
4      nonexistent          1.1          93.994          -36.4      4.857   
...            ...          ...             ...            ...        ...   
41183  nonexistent         -1.1          94.767          -50.8      1.028   
41184  nonexistent         -1.1          94.767          -50.8      1.028   
41185  nonexistent         -1.1          94.767          -50.8      1.028   
41186  nonexistent         -1.1          94.767          -50.8      1.028   
41187      failure         -1.1          94.767          -50.8      1.028   

       nr.employed    y  
0           5191.0   no  
1           5191.0   no  
2           5191.0   no  
3           5191.0   no  
4           5191.0   no  
...            ...  ...  
41183       4963.6  yes  
41184       4963.6   no  
41185       4963.6   no  
41186       4963.6  yes  
41187       4963.6   no  

[41188 rows x 21 columns]

## Experiment 3: 
The training and test DataFrames will need some significant preprocessing before they can be used:

a. Several of the features are categorical variables and will need to be turned into numbers before they can be used by ML algorithms. The simplest way to accomplish this is to use dummy coding using the Pandas get_dummies() method. Note: Some algorithms (e.g. logistic regression) have problems with collinear features. If you use one-hot encoding, one dummy variable will be a linear combination of the other dummy variables, so be sure to pass drop_first=True.

b. Two features should be removed from the dataset:
i. Per bank-additional-names.txt, the duration “should be discarded if the intention is to have a realistic predictive model.”
ii.The feature y (or y_yes after dummy coding) is the target.

c. Some algorithms (e.g. KNN and SVM) require non-categorical features to be standardized. 


### The Result of Experiment 3 (a): 

In [4]:
dfNewTrainingSet = pd.get_dummies(dfTrainingSet, columns=['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'day_of_week', 'poutcome', 'y'], drop_first=True)
dfNewTrainingSet

age  duration  campaign  pdays  previous  emp.var.rate  cons.price.idx  \
0      30       487         2    999         0          -1.8          92.893   
1      39       346         4    999         0           1.1          93.994   
2      25       227         1    999         0           1.4          94.465   
3      38        17         3    999         0           1.4          94.465   
4      47        58         1    999         0          -0.1          93.200   
...   ...       ...       ...    ...       ...           ...             ...   
4114   30        53         1    999         0           1.4          93.918   
4115   39       219         1    999         0           1.4          93.918   
4116   27        64         2    999         1          -1.8          92.893   
4117   58       528         1    999         0           1.4          93.444   
4118   34       175         1    999         0          -0.1          93.200   

      cons.conf.idx  euribor3m  nr.employed  ...  month_nov  month_oct  \
0             -46.2      1.313       5099.1  ...          0          0   
1             -36.4      4.855       5191.0  ...          0          0   
2             -41.8      4.962       5228.1  ...          0          0   
3             -41.8      4.959       5228.1  ...          0          0   
4             -42.0      4.191       5195.8  ...          1          0   
...             ...        ...          ...  ...        ...        ...   
4114          -42.7      4.958       5228.1  ...          0          0   
4115          -42.7      4.959       5228.1  ...          0          0   
4116          -46.2      1.354       5099.1  ...          0          0   
4117          -36.1      4.966       5228.1  ...          0          0   
4118          -42.0      4.120       5195.8  ...          1          0   

      month_sep  day_of_week_mon  day_of_week_thu  day_of_week_tue  \
0             0                0                0                0   
1             0                0                0                0   
2             0                0                0                0   
3             0                0                0                0   
4             0                1                0                0   
...         ...              ...              ...              ...   
4114          0                0                1                0   
4115          0                0                0                0   
4116          0                1                0                0   
4117          0                0                0                0   
4118          0                0                0                0   

      day_of_week_wed  poutcome_nonexistent  poutcome_success  y_yes  
0                   0                     1                 0      0  
1                   0                     1                 0      0  
2                   1                     1                 0      0  
3                   0                     1                 0      0  
4                   0                     1                 0      0  
...               ...                   ...               ...    ...  
4114                0                     1                 0      0  
4115                0                     1                 0      0  
4116                0                     0                 0      0  
4117                0                     1                 0      0  
4118                1                     1                 0      0  

[4119 rows x 54 columns]

In [5]:
dfNewTestSet = pd.get_dummies(dfTestSet, columns=['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'day_of_week', 'poutcome', 'y'], drop_first=True)
dfNewTestSet

age  duration  campaign  pdays  previous  emp.var.rate  cons.price.idx  \
0       56       261         1    999         0           1.1          93.994   
1       57       149         1    999         0           1.1          93.994   
2       37       226         1    999         0           1.1          93.994   
3       40       151         1    999         0           1.1          93.994   
4       56       307         1    999         0           1.1          93.994   
...    ...       ...       ...    ...       ...           ...             ...   
41183   73       334         1    999         0          -1.1          94.767   
41184   46       383         1    999         0          -1.1          94.767   
41185   56       189         2    999         0          -1.1          94.767   
41186   44       442         1    999         0          -1.1          94.767   
41187   74       239         3    999         1          -1.1          94.767   

       cons.conf.idx  euribor3m  nr.employed  ...  month_nov  month_oct  \
0              -36.4      4.857       5191.0  ...          0          0   
1              -36.4      4.857       5191.0  ...          0          0   
2              -36.4      4.857       5191.0  ...          0          0   
3              -36.4      4.857       5191.0  ...          0          0   
4              -36.4      4.857       5191.0  ...          0          0   
...              ...        ...          ...  ...        ...        ...   
41183          -50.8      1.028       4963.6  ...          1          0   
41184          -50.8      1.028       4963.6  ...          1          0   
41185          -50.8      1.028       4963.6  ...          1          0   
41186          -50.8      1.028       4963.6  ...          1          0   
41187          -50.8      1.028       4963.6  ...          1          0   

       month_sep  day_of_week_mon  day_of_week_thu  day_of_week_tue  \
0              0                1                0                0   
1              0                1                0                0   
2              0                1                0                0   
3              0                1                0                0   
4              0                1                0                0   
...          ...              ...              ...              ...   
41183          0                0                0                0   
41184          0                0                0                0   
41185          0                0                0                0   
41186          0                0                0                0   
41187          0                0                0                0   

       day_of_week_wed  poutcome_nonexistent  poutcome_success  y_yes  
0                    0                     1                 0      0  
1                    0                     1                 0      0  
2                    0                     1                 0      0  
3                    0                     1                 0      0  
4                    0                     1                 0      0  
...                ...                   ...               ...    ...  
41183                0                     1                 0      1  
41184                0                     1                 0      0  
41185                0                     1                 0      0  
41186                0                     1                 0      1  
41187                0                     0                 0      0  

[41188 rows x 54 columns]

### The Result of Experiment 3 (b): 

In [6]:
dfNewTrainingSet.drop('duration', inplace=True, axis=1)
dfNewTrainingSet

age  campaign  pdays  previous  emp.var.rate  cons.price.idx  \
0      30         2    999         0          -1.8          92.893   
1      39         4    999         0           1.1          93.994   
2      25         1    999         0           1.4          94.465   
3      38         3    999         0           1.4          94.465   
4      47         1    999         0          -0.1          93.200   
...   ...       ...    ...       ...           ...             ...   
4114   30         1    999         0           1.4          93.918   
4115   39         1    999         0           1.4          93.918   
4116   27         2    999         1          -1.8          92.893   
4117   58         1    999         0           1.4          93.444   
4118   34         1    999         0          -0.1          93.200   

      cons.conf.idx  euribor3m  nr.employed  job_blue-collar  ...  month_nov  \
0             -46.2      1.313       5099.1                1  ...          0   
1             -36.4      4.855       5191.0                0  ...          0   
2             -41.8      4.962       5228.1                0  ...          0   
3             -41.8      4.959       5228.1                0  ...          0   
4             -42.0      4.191       5195.8                0  ...          1   
...             ...        ...          ...              ...  ...        ...   
4114          -42.7      4.958       5228.1                0  ...          0   
4115          -42.7      4.959       5228.1                0  ...          0   
4116          -46.2      1.354       5099.1                0  ...          0   
4117          -36.1      4.966       5228.1                0  ...          0   
4118          -42.0      4.120       5195.8                0  ...          1   

      month_oct  month_sep  day_of_week_mon  day_of_week_thu  day_of_week_tue  \
0             0          0                0                0                0   
1             0          0                0                0                0   
2             0          0                0                0                0   
3             0          0                0                0                0   
4             0          0                1                0                0   
...         ...        ...              ...              ...              ...   
4114          0          0                0                1                0   
4115          0          0                0                0                0   
4116          0          0                1                0                0   
4117          0          0                0                0                0   
4118          0          0                0                0                0   

      day_of_week_wed  poutcome_nonexistent  poutcome_success  y_yes  
0                   0                     1                 0      0  
1                   0                     1                 0      0  
2                   1                     1                 0      0  
3                   0                     1                 0      0  
4                   0                     1                 0      0  
...               ...                   ...               ...    ...  
4114                0                     1                 0      0  
4115                0                     1                 0      0  
4116                0                     0                 0      0  
4117                0                     1                 0      0  
4118                1                     1                 0      0  

[4119 rows x 53 columns]

In [7]:
# Storing the feature y_yes as the target before removing it from the dataframe
targetTraining = dfNewTrainingSet['y_yes']

In [8]:
dfNewTrainingSet.drop('y_yes', inplace=True, axis=1)
dfNewTrainingSet

age  campaign  pdays  previous  emp.var.rate  cons.price.idx  \
0      30         2    999         0          -1.8          92.893   
1      39         4    999         0           1.1          93.994   
2      25         1    999         0           1.4          94.465   
3      38         3    999         0           1.4          94.465   
4      47         1    999         0          -0.1          93.200   
...   ...       ...    ...       ...           ...             ...   
4114   30         1    999         0           1.4          93.918   
4115   39         1    999         0           1.4          93.918   
4116   27         2    999         1          -1.8          92.893   
4117   58         1    999         0           1.4          93.444   
4118   34         1    999         0          -0.1          93.200   

      cons.conf.idx  euribor3m  nr.employed  job_blue-collar  ...  month_may  \
0             -46.2      1.313       5099.1                1  ...          1   
1             -36.4      4.855       5191.0                0  ...          1   
2             -41.8      4.962       5228.1                0  ...          0   
3             -41.8      4.959       5228.1                0  ...          0   
4             -42.0      4.191       5195.8                0  ...          0   
...             ...        ...          ...              ...  ...        ...   
4114          -42.7      4.958       5228.1                0  ...          0   
4115          -42.7      4.959       5228.1                0  ...          0   
4116          -46.2      1.354       5099.1                0  ...          1   
4117          -36.1      4.966       5228.1                0  ...          0   
4118          -42.0      4.120       5195.8                0  ...          0   

      month_nov  month_oct  month_sep  day_of_week_mon  day_of_week_thu  \
0             0          0          0                0                0   
1             0          0          0                0                0   
2             0          0          0                0                0   
3             0          0          0                0                0   
4             1          0          0                1                0   
...         ...        ...        ...              ...              ...   
4114          0          0          0                0                1   
4115          0          0          0                0                0   
4116          0          0          0                1                0   
4117          0          0          0                0                0   
4118          1          0          0                0                0   

      day_of_week_tue  day_of_week_wed  poutcome_nonexistent  poutcome_success  
0                   0                0                     1                 0  
1                   0                0                     1                 0  
2                   0                1                     1                 0  
3                   0                0                     1                 0  
4                   0                0                     1                 0  
...               ...              ...                   ...               ...  
4114                0                0                     1                 0  
4115                0                0                     1                 0  
4116                0                0                     0                 0  
4117                0                0                     1                 0  
4118                0                1                     1                 0  

[4119 rows x 52 columns]

In [9]:
dfNewTestSet.drop('duration', inplace=True, axis=1)
dfNewTestSet

age  campaign  pdays  previous  emp.var.rate  cons.price.idx  \
0       56         1    999         0           1.1          93.994   
1       57         1    999         0           1.1          93.994   
2       37         1    999         0           1.1          93.994   
3       40         1    999         0           1.1          93.994   
4       56         1    999         0           1.1          93.994   
...    ...       ...    ...       ...           ...             ...   
41183   73         1    999         0          -1.1          94.767   
41184   46         1    999         0          -1.1          94.767   
41185   56         2    999         0          -1.1          94.767   
41186   44         1    999         0          -1.1          94.767   
41187   74         3    999         1          -1.1          94.767   

       cons.conf.idx  euribor3m  nr.employed  job_blue-collar  ...  month_nov  \
0              -36.4      4.857       5191.0                0  ...          0   
1              -36.4      4.857       5191.0                0  ...          0   
2              -36.4      4.857       5191.0                0  ...          0   
3              -36.4      4.857       5191.0                0  ...          0   
4              -36.4      4.857       5191.0                0  ...          0   
...              ...        ...          ...              ...  ...        ...   
41183          -50.8      1.028       4963.6                0  ...          1   
41184          -50.8      1.028       4963.6                1  ...          1   
41185          -50.8      1.028       4963.6                0  ...          1   
41186          -50.8      1.028       4963.6                0  ...          1   
41187          -50.8      1.028       4963.6                0  ...          1   

       month_oct  month_sep  day_of_week_mon  day_of_week_thu  \
0              0          0                1                0   
1              0          0                1                0   
2              0          0                1                0   
3              0          0                1                0   
4              0          0                1                0   
...          ...        ...              ...              ...   
41183          0          0                0                0   
41184          0          0                0                0   
41185          0          0                0                0   
41186          0          0                0                0   
41187          0          0                0                0   

       day_of_week_tue  day_of_week_wed  poutcome_nonexistent  \
0                    0                0                     1   
1                    0                0                     1   
2                    0                0                     1   
3                    0                0                     1   
4                    0                0                     1   
...                ...              ...                   ...   
41183                0                0                     1   
41184                0                0                     1   
41185                0                0                     1   
41186                0                0                     1   
41187                0                0                     0   

       poutcome_success  y_yes  
0                     0      0  
1                     0      0  
2                     0      0  
3                     0      0  
4                     0      0  
...                 ...    ...  
41183                 0      1  
41184                 0      0  
41185                 0      0  
41186                 0      1  
41187                 0      0  

[41188 rows x 53 columns]

In [10]:
# Storing the feature y_yes as the target before removing it from the dataframe
targetTest = dfNewTestSet['y_yes']

In [11]:
dfNewTestSet.drop('y_yes', inplace=True, axis=1)

In [12]:
dfNewTestSet

age  campaign  pdays  previous  emp.var.rate  cons.price.idx  \
0       56         1    999         0           1.1          93.994   
1       57         1    999         0           1.1          93.994   
2       37         1    999         0           1.1          93.994   
3       40         1    999         0           1.1          93.994   
4       56         1    999         0           1.1          93.994   
...    ...       ...    ...       ...           ...             ...   
41183   73         1    999         0          -1.1          94.767   
41184   46         1    999         0          -1.1          94.767   
41185   56         2    999         0          -1.1          94.767   
41186   44         1    999         0          -1.1          94.767   
41187   74         3    999         1          -1.1          94.767   

       cons.conf.idx  euribor3m  nr.employed  job_blue-collar  ...  month_may  \
0              -36.4      4.857       5191.0                0  ...          1   
1              -36.4      4.857       5191.0                0  ...          1   
2              -36.4      4.857       5191.0                0  ...          1   
3              -36.4      4.857       5191.0                0  ...          1   
4              -36.4      4.857       5191.0                0  ...          1   
...              ...        ...          ...              ...  ...        ...   
41183          -50.8      1.028       4963.6                0  ...          0   
41184          -50.8      1.028       4963.6                1  ...          0   
41185          -50.8      1.028       4963.6                0  ...          0   
41186          -50.8      1.028       4963.6                0  ...          0   
41187          -50.8      1.028       4963.6                0  ...          0   

       month_nov  month_oct  month_sep  day_of_week_mon  day_of_week_thu  \
0              0          0          0                1                0   
1              0          0          0                1                0   
2              0          0          0                1                0   
3              0          0          0                1                0   
4              0          0          0                1                0   
...          ...        ...        ...              ...              ...   
41183          1          0          0                0                0   
41184          1          0          0                0                0   
41185          1          0          0                0                0   
41186          1          0          0                0                0   
41187          1          0          0                0                0   

       day_of_week_tue  day_of_week_wed  poutcome_nonexistent  \
0                    0                0                     1   
1                    0                0                     1   
2                    0                0                     1   
3                    0                0                     1   
4                    0                0                     1   
...                ...              ...                   ...   
41183                0                0                     1   
41184                0                0                     1   
41185                0                0                     1   
41186                0                0                     1   
41187                0                0                     0   

       poutcome_success  
0                     0  
1                     0  
2                     0  
3                     0  
4                     0  
...                 ...  
41183                 0  
41184                 0  
41185                 0  
41186                 0  
41187                 0  

[41188 rows x 52 columns]

### The Result of Experiment 3 (c): 

In [13]:
scaler= StandardScaler()
# standardization of feture variables which are now all non-categorical features
scaled_TrainingSet = scaler.fit_transform(dfNewTrainingSet)
scaled_TestSet = scaler.fit_transform(dfNewTestSet)

## Experiment 4: 
Fit Naive Bayes, KNN, and SVM classifiers to the training set, then score each classifier on the test set. Which classifier has the highest accuracy?

In [14]:
gnbTraining = GaussianNB().fit(scaled_TrainingSet, targetTraining)
gnbTraining.score(scaled_TestSet,targetTest)

0.8873458288821987

In [15]:
neighTraining = KNeighborsClassifier(n_neighbors=3).fit(scaled_TrainingSet, targetTraining)
neighTraining.score(scaled_TestSet,targetTest)

0.88671457706128

In [16]:
clf = make_pipeline(StandardScaler(), SVC(gamma='auto'))
svmTraining = clf.fit(scaled_TrainingSet, targetTraining)

svmTraining.score(scaled_TestSet,targetTest)

0.8981985044187627

## The result of Experiment 4: 

According to the scores I got above, Naive Bayes, KNN, and SVM classifiers scored 88.73%, 88.67%, and 89.82% respectively. Therefore, SVM classifier has the highest accuracy among the other classifiers. 

## Experiment 5: 

How many values in the training set have y_yes = 0, and how many have y_yes = 1? What would be the accuracy if we simply assumed that no customer ever subscribed to the product?

In [17]:
targetTraining.value_counts()

0    3668
1     451
Name: y_yes, dtype: int64

## The result of Experiment 5: 

There are 3668 values out of 4119 that have y_yes = 0 and 451 values out of 4119 that have y_yes = 1 in the training set. When no customer ever subscribed to the product is assumed, the accuracy would measure the portion that is only in one class which is y_yes = 0. There would be no machine learning happening this way, not a good classiifer. 

## Experiment 6: 

Use np.zeros_like() to create a target vector representing the output of the “dumb” classifier of experiment (5), then create a confusion matrix and find its AUC.

In [18]:
from sklearn.metrics import confusion_matrix

# y_test
y_true = np.zeros_like(targetTest)

In [19]:
#prediction on the fitted model
y_pred = svmTraining.predict(scaled_TestSet)

confusion_matrix(y_true, y_pred)

array([[40031,  1157],
       [    0,     0]], dtype=int64)

In [20]:
#prediction on the fitted model
y_pred = neighTraining.predict(scaled_TestSet)

confusion_matrix(y_true, y_pred)

array([[38712,  2476],
       [    0,     0]], dtype=int64)

## The result of Experiment 6: 

AUC: I got the row of zeros in both of the confusion matrices using KNN and SVM classifiers, meaning that there is no negative predictions. In class, we learned that we are not supposed to get rows or columns of zero in the confusion matrix. Since we created a target vector of all zeros as the output of the dumb classifier from the experiment (5), AUC cannot be computed here since we have only one class to work with.

## Experiment 7: 
Create a confusion matrix and find the AUC for each of the classifiers of experiment (4). Is the best classifier the one with the highest accuracy?

In [21]:
#prediction on the fitted model from Naive Bayes Classifier
y_pred1 = gnbTraining.predict(scaled_TestSet)

confusion_matrix(targetTest, y_pred1)

array([[36548,     0],
       [ 4640,     0]], dtype=int64)

In [22]:
from sklearn.metrics import roc_auc_score

roc_auc_score(targetTest, y_pred1)

0.5

In [23]:
#prediction on the fitted model from KNN Classifier
y_pred2 = neighTraining.predict(scaled_TestSet)

confusion_matrix(targetTest, y_pred2)

array([[35297,  1251],
       [ 3415,  1225]], dtype=int64)

In [24]:
roc_auc_score(targetTest, y_pred2)

0.6148898307563413

In [25]:
#prediction on the fitted model from SVM  Classifier
y_pred3 = svmTraining.predict(scaled_TestSet)

confusion_matrix(targetTest, y_pred3)

array([[36193,   355],
       [ 3838,   802]], dtype=int64)

In [26]:
roc_auc_score(targetTest, y_pred3)

0.5815657868914946

## The result of Experiment 7: 
No, the best classifier in this experiment with an AUC score of 61.49% is KNN classifier, but SVM classifier had the highest accuracy of 89.82% score from experiment (4). Therefore, this indicates an imbalanced dataset which leads to a high accuracy with a misleading mesure.

## Experiment 8: 
One of the easiest ways to deal with an unbalanced dataset is random oversampling. This can be done with an imblearn.over_sampling.RandomOverSampler object. Use fit_resample() to generate a balanced training set. To make sure that your results are reproducible, pass random_state=(2021-4-22).

In [27]:
from imblearn.over_sampling import RandomOverSampler

ros = RandomOverSampler(random_state=2021-4-22)

resampledTraining, resampledTargetTraining = ros.fit_resample(scaled_TrainingSet, targetTraining)

## Experiment 9: 
Repeat experiments (4) and (7) on the balanced training set of experiment (8). Which classifier performs the best, and how much better is its performance?

In [28]:
gnbResampled = GaussianNB().fit(resampledTraining, resampledTargetTraining)
gnbResampled.score(scaled_TestSet,targetTest)

0.8873458288821987

In [29]:
#prediction on the fitted model from Naive Bayes Classifier
y_pred1 = gnbResampled.predict(scaled_TestSet)

confusion_matrix(targetTest, y_pred1)

array([[36548,     0],
       [ 4640,     0]], dtype=int64)

In [30]:
roc_auc_score(targetTest, y_pred1)

0.5

In [31]:
neighResampled = KNeighborsClassifier(n_neighbors=3).fit(resampledTraining, resampledTargetTraining)
neighResampled.score(scaled_TestSet,targetTest)

0.8088035350101972

In [32]:
#prediction on the fitted model from KNN Classifier
y_pred2 = neighResampled.predict(scaled_TestSet)

confusion_matrix(targetTest, y_pred2)

array([[30996,  5552],
       [ 2323,  2317]], dtype=int64)

In [33]:
roc_auc_score(targetTest, y_pred2)

0.6737218155246006

In [34]:
svmResampled = clf.fit(resampledTraining, resampledTargetTraining)
svmResampled.score(scaled_TestSet,targetTest)

0.8328153831212974

In [35]:
#prediction on the fitted model from SVM  Classifier
y_pred3 = svmResampled.predict(scaled_TestSet)

confusion_matrix(targetTest, y_pred3)

array([[31791,  4757],
       [ 2129,  2511]], dtype=int64)

In [36]:
roc_auc_score(targetTest, y_pred3)

0.7055030960701657

## The result of Experiment 9: 
After generating a balanced training set, SVM classifier performed better according to its AUC score of 70.55%. SVM classifier had an AUC score of 58.16% before generating a balanced training set, so its performance improved by 12.39% using a balanced training set.